File: src/Q9/jubin/Solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        valid_def = {'(': ')', '[': ']', '{': '}'}

        check_stack = []
        for item in s:
            if valid_def.get(item):
                check_stack.append(item)
            else:
                if not check_stack:
                    return False
                if valid_def.get(check_stack.pop()) != item:
                    return False
        else:
            if check_stack:
                return False
        return True


if __name__ == '__main__':
    test = Solution()

    print(test.isValid("()"))
    print(test.isValid(")("))
    print(test.isValid("()[]{}"))
    print(test.isValid("(]"))
    print(test.isValid("([)]"))
    print(test.isValid("{[]}"))
    print(test.isValid("(])"))

File: src/Q9/jubin/Solution.java
package Q9.jubin;

import java.util.HashMap;
import java.util.Stack;

public class Solution {
    public static boolean isValid(String s) {
        HashMap<Character, Character> valid_def = new HashMap<>();
        Stack<Character> check_stack = new Stack<>();

        valid_def.put('(', ')');
        valid_def.put('{', '}');
        valid_def.put('[', ']');

        for(char char_item : s.toCharArray()){
            if (valid_def.get(char_item) != null){
                check_stack.push(char_item);
            }else{
                if(check_stack.empty()){
                    return false;
                }else{
                    if(valid_def.get(check_stack.pop()) != char_item){
                        return false;
                    }
                }
            }
        }

        return check_stack.empty();
    }

    public static void main(String[] args) {
        System.out.println(isValid("()"));
        System.out.println(isValid(")("));
        System.out.println(isValid("()[]{}"));
        System.out.println(isValid("(]"));
        System.out.println(isValid("([)]"));
        System.out.println(isValid("{[]}"));
        System.out.println(isValid("(])"));

    }
}
